File: src/main/java/kh/farrukh/aoc2024/day2/Day2A.java
package kh.farrukh.aoc2024.day2;

import static java.lang.Math.abs;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Day2A {

  public static void main(String[] args) {
    var sc = new Scanner(System.in);

    List<List<Integer>> reports = new ArrayList<>();

    while (sc.hasNextLine()) {
      var line = sc.nextLine();
      if (line.isEmpty()) {
        break;
      }
      var report = new ArrayList<Integer>();
      for (var s : line.split(" ")) {
        report.add(Integer.parseInt(s));
      }
      reports.add(report);
    }

    var monotonous = 0;

    for (var report : reports) {
      var isMono = true;
      var isIncreasing = report.get(0) < report.get(1);

      var diffFirst = abs(report.get(0) - report.get(1));
      if (diffFirst < 1 || diffFirst > 3) {
        isMono = false;
      } else {
        for (int i = 1; i < report.size() - 1; i++) {
          if (isIncreasing && report.get(i) > report.get(i + 1)) {
            isMono = false;
            break;
          }
          if (!isIncreasing && report.get(i) < report.get(i + 1)) {
            isMono = false;
            break;
          }
          var diff = abs(report.get(i) - report.get(i + 1));
          if (diff < 1 || diff > 3) {
            isMono = false;
            break;
          }
        }
      }

      if (isMono) {
        monotonous++;
      }
    }

    System.out.println(monotonous);

  }

}

File: src/main/java/kh/farrukh/aoc2024/day2/Day2A.kt
package kh.farrukh.aoc2024.day2

import java.util.*
import kotlin.math.abs

fun main() {
    val sc = Scanner(System.`in`)

    val reports = mutableListOf<List<Int>>()

    while (sc.hasNextLine()) {
        val line = sc.nextLine()
        if (line.isBlank()) {
            break
        }
        val report = line.split(" ").map { it.toInt() }
        reports.add(report)
    }

    var validCount = 0

    for (report in reports) {
        var isValidReport = true
        val isIncreasing = report[1] > report[0]

        for (i in 1 until report.size) {
            val current = report[i]
            val prev = report[i - 1]

            if (!check(current, prev, isIncreasing)) {
                isValidReport = false
                break
            }
        }

        if (isValidReport) {
            validCount++
        }
    }

    println(validCount)
}

fun check(
    current: Int,
    prev: Int,
    isIncreasing: Boolean
): Boolean {
    if (isIncreasing && current < prev) {
        return false
    }

    if (!isIncreasing && current > prev) {
        return false
    }

    val diff = abs(current - prev)
    return diff in 1..3
}